### PR TITLE
Transformer v1

### PR DESCRIPTION
**Pull request by lolleko**
This PR adds complete transformer support.
All transform functions are implemented,
but we still need to move a lot of functionality from the Transpiler to the Transformer.

WIP since tests are failing and some polishing is required.

### CODE REVIEW

File: test/translation/builder.spec.ts [hunk 1]
```diff
@@ -1,47 +1,48 @@
-import { Expect, Test, TestCases } from "alsatian";
+import { Expect, FocusTest, Test, TestCases } from "alsatian";
```
Comment (Perryvw) [inline test/translation/builder.spec.ts]: This should probably not be comitted

File: test/runner.ts [hunk 2]
```diff
@@ -23,7 +23,7 @@ fs.copyFileSync(
 testRunner.outputStream
           // this will use alsatian's default output if you remove this
           // you'll get TAP or you can add your favourite TAP reporter in it's place
-          .pipe(TapBark.create().getPipeable())
+          // .pipe(TapBark.create().getPipeable())
```
Comment (Perryvw) [inline test/runner.ts]: Should probably revert this before merging.

File: test/translation/builder.spec.ts [hunk 3]
```diff
@@ -1,47 +1,48 @@
...
-    return b.toString().trim().split("\r\n").join("\n")
+    return b.toString().trim().split("\r\n").join("\n");
 }
 
 export class FileTests {
 
     @TestCases(files)
     @Test("Transformation Tests")
-    public transformationTests(tsFile: string, luaFile:string) {
+    @FocusTest
```
Comment (Perryvw) [inline test/translation/builder.spec.ts]: Are you sure this should be comitted?

File: src/TSHelper.ts [hunk 4]
```diff
@@ -56,6 +56,16 @@ export class TSHelper {
         return false;
     }
 
+    public static isIdentifierExported(
+        identifier: ts.Identifier, scope: ts.ModuleDeclaration | ts.SourceFile, checker: ts.TypeChecker): boolean {
```
Comment (Perryvw) [inline src/TSHelper.ts]: formatting of this function header looks pretty strange.

File: src/Transformer.ts [hunk 5]
```diff
@@ -63,6 +68,78 @@ export class LuaTransformer {
...
+            case ts.SyntaxKind.ModuleBlock:
+                return this.visitModuleBlock(node as ts.ModuleBlock);
+            // EOF TOKEN
+            case ts.SyntaxKind.EndOfFileToken:
+                return this.visitEndOfFileToken(node as ts.EndOfFileToken);
+            default:
+                throw TSTLErrors.UnsupportedKind("Node", node.kind, node);
+        }
+    }
+    public visitStatement(node: ts.Statement): ts.Statement {
```
Comment (Perryvw) [inline src/Transformer.ts]: We should probably consistently add a white-line between methods

File: src/Transformer.ts [hunk 6]
```diff
@@ -63,6 +68,78 @@ export class LuaTransformer {
...
+            case ts.SyntaxKind.DoStatement:
+            case ts.SyntaxKind.ForStatement:
+            case ts.SyntaxKind.ForOfStatement:
+            case ts.SyntaxKind.ForInStatement:
+            case ts.SyntaxKind.SwitchStatement:
+            case ts.SyntaxKind.BreakStatement:
+            case ts.SyntaxKind.TryStatement:
+            case ts.SyntaxKind.ThrowStatement:
+            case ts.SyntaxKind.ContinueStatement:
+            case ts.SyntaxKind.EmptyStatement:
```
Comment (Perryvw) [inline src/Transformer.ts]: If you want to split like this, is there not some kind of `ts.IsStatement(node)`?

Comment (lolleko) [reply src/Transformer.ts]: There is not, maybe we can reorganise it to look a bit nicer later

File: src/Transformer.ts (hunk 5 above)
Comment (lolleko) [reply src/Transformer.ts]: Added in 3e791a4

**Pull request merged**